src/Domain/Services/CheckoutFields.php

```diff
@@ -0,0 +1,653 @@
+	 */
+	public function get_additional_fields() {
+		return $this->additional_fields;
+	}
+
+	/**
+	 * Update the default locale with additional fields without country limitations.
+	 *
+	 * @param array $locale The locale to update.
+	 * @return mixed
+	 */
+	public function update_default_locale_with_fields( $locale ) {
+		foreach ( $this->fields_locations['address'] as $field_id => $additional_field ) {
+			if ( empty( $locale[ $field_id ] ) ) {
+				$locale[ $field_id ] = $additional_field;
+			}
+		}
+		return $locale;
+	}
+
+	/**
+	 * Returns an array of fields keys for the address group.
+	 *
+	 * @return array An array of fields keys.
+	 */
+	public function get_address_fields_keys() {
+		return $this->fields_locations['address'];
+	}
+
+	/**
+	 * Returns an array of fields keys for the contact group.
+	 *
+	 * @return array An array of fields keys.
+	 */
+	public function get_contact_fields_keys() {
+		return $this->fields_locations['contact'];
+	}
+
+	/**
+	 * Returns an array of fields keys for the additional area group.
+	 *
+	 * @return array An array of fields keys.
+	 */
+	public function get_additional_fields_keys() {
+		return $this->fields_locations['additional'];
+	}
+
+	/**
+	 * Returns an array of fields for a given group.
+	 *
+	 * @param string $location The location to get fields for (address|contact|additional).
+	 *
+	 * @return array An array of fields.
+	 */
+	public function get_fields_for_location( $location ) {
+		if ( in_array( $location, array_keys( $this->fields_locations ), true ) ) {
+			return $this->fields_locations[ $location ];
+		}
+	}
+
+	/**
+	 * Validates a field value for a given group.
+	 *
+	 * @param string $key The field key.
+	 * @param mixed  $value The field value.
+	 * @param string $location The location to validate the field for (address|contact|additional).
+	 *
+	 * TODO: we might not need the location param here.
+	 *
+	 * @return true|\WP_Error True if the field is valid, a WP_Error otherwise.
+	 */
+	public function validate_field_for_location( $key, $value, $location ) {
+		if ( ! $this->is_field( $key ) ) {
+			// translators: %s field key.
+			return new \WP_Error( 'woocommerce_blocks_checkout_field_invalid', \sprintf( __( 'The field %s is invalid.', 'woo-gutenberg-products-block' ), $key ) );
+		}
+
+		if ( ! in_array( $key, $this->fields_locations[ $location ], true ) ) {
+			// translators: %1$s field key, %2$s location.
+			return new \WP_Error( 'woocommerce_blocks_checkout_field_invalid_location', \sprintf( __( 'The field %1$s is invalid for the location %2$s.', 'woo-gutenberg-products-block' ), $key, $location ) );
+		}
+
+		$field = $this->additional_fields[ $key ];
+		if ( ! empty( $field['required'] ) && empty( $value ) ) {
+			// translators: %s field key.
+			return new \WP_Error( 'woocommerce_blocks_checkout_field_required', \sprintf( __( 'The field %s is required.', 'woo-gutenberg-products-block' ), $key ) );
+		}
+
+		return true;
+	}
```

Comment:
Validating a key for a location just make sure you're passing the field in the right place, so not passing an address field in additional. It also checks if the field is required and set.

In the future, this will also include developer provided validation function.